api/envoy/admin/v2alpha/clusters.proto

```diff
@@ -0,0 +1,93 @@
+import "envoy/type/percent.proto";
+
+// [#protodoc-title: Clusters]
+
+// Admin endpoint uses this wrapper for `/clusters` to display cluster status information.
+// See :ref:`/clusters <operations_admin_interface_clusters>` for more information.
+message Clusters {
+  // Mapping from cluster name to each cluster's status.
+  repeated ClusterStatus cluster_statuses = 1;
+}
+
+// Details an individual cluster's current status.
+message ClusterStatus {
+  // Name of the cluster.
+  string name = 1;
+
+  // General outlier statistics if installed for this cluster.
+  OutlierInfo outlier_info = 2;
+
+  // Denotes whether this cluster was added via API or configured statically.
+  bool added_via_api = 3;
+
+  // Mapping from host address to the host's current status.
+  repeated HostStatus host_statuses = 4;
+}
+
+// :ref:`Cluster outlier detection <arch_overview_outlier_detection>` statistics. -1 for any of
+// these statistics denotes that there was not enough data to compute it.
+message OutlierInfo {
+  // The average success rate of the hosts in the Detector for the last aggregation interval.
+  //
+  // Note: the message will be omitted if there were not enough hosts with enough request volume to
+  // proceed with success rate based outlier ejection.
+  envoy.type.Percent success_rate_average = 1;
+
+  // The success rate threshold used in the last interval. The threshold is used to eject hosts
+  // based on their success rate.
+  //
+  // Note: the message will be omitted if there were not enough hosts with enough request volume to
+  // proceed with success rate based outlier ejection. Since this value can technically be negative
+  // due to the way that it's computed, negative values will also be represented by the absence of
+  // this field because a negative value implies that there was no threshold for that interval.
+  envoy.type.Percent success_rate_ejection_threshold = 2;
+}
+
+// Current state of a particular host.
+message HostStatus {
+  // Address of this host.
+  envoy.api.v2.core.Address address = 1;
+
+  // Mapping from the name of the statistic to the current value.
+  map<string, int64> stats = 2;
+
+  // The host's current health status.
+  HostHealthStatus health_status = 3;
+
+  // Configured load balancing weight for this host.
+  uint64 weight = 4;
+
+  // Configured locality for the host.
+  envoy.api.v2.core.Locality locality = 5;
+
+  // Denotes whether this is configured as a canary host or not.
+  bool canary = 6;
+
+  // Request success rate for this host over the last calculated interval.
+  //
+  // Note: the message will not be present if host did not have enough request volume to calculate
+  // success rate or the cluster did not have enough hosts to run through success rate outlier
+  // ejection.
+  envoy.type.Percent success_rate = 7;
+}
+
+// Health status for a host.
+message HostHealthStatus {
+  // The host is currently marked as healthy.
+  bool healthy = 1;
```

Comment:
I think we can simplify this to just indicating when a host is unhealthy, and then when there are no bits set saying it's unhealthy, it's considered healthy. I wonder how this relates to https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/health_check.proto#L182.. it seems we should have the ability to also indicate draining? Should we have a single health status message?

Reply:
As for the former, I thought about that.  My only hesitation is from a usability perspective.  The highest order bit here is whether the host is healthy or not, but that would be the hardest for the user to determine if we were to eliminate the redundant `healthy` bool.  You could probably get a little clever by omitting this message or a submessage for a healthy host to create an artificial way to signal a healthy host, but that might be more trouble than it's worth.  WDYT?

As for how this relates to the EDS health status, that is reduced to a bool here IIUC - `failed_eds_health_check`.  We could make this bool an enum to represent the full range of EDS health states rather than just healthy or unhealthy.

Reply:
I think we don't have to spend too much time making these "usable", in that the APIs are intended for machine parsing first. So, I think we can drop the healthy status. The EDS health status is really a reflection of other health status possibilities. We might for example, via HDS, discover that a host is draining independent to EDS. So, I think we should be able to convey these possibilities. I would argue that we probably just want a single health status proto and it should be https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/health_check.proto#L182, we should make it work if it's not expressive enough for the requiremetns here (e.g. indicating outlier detection).